File: moogli/bin/src/view/Viewer.cpp
#include "view/Viewer.hpp"

Viewer::Viewer( QWidget * parent
              , const QGLWidget * share_widget
              , Qt::WindowFlags f
              ): QGLWidget( parent
                          , share_widget
                          , f)
               , _viewer(new osgViewer::CompositeViewer())
               , _shapes(new osg::PositionAttitudeTransform())
{

    QGLWidget::setFormat(QGLFormat(QGL::SampleBuffers));

    osg::ref_ptr<osg::GraphicsContext::Traits> traits = new osg::GraphicsContext::Traits();
    traits -> x = this -> x();
    traits -> y = this -> y();
    traits -> width = this -> width();
    traits -> height = this -> height();
    traits -> windowDecoration = false;
    traits -> doubleBuffer = true;
    traits -> sharedContext = 0;
    traits -> samples = 4;
    _graphics_window = new osgViewer::GraphicsWindowEmbedded(traits.get());

    _shapes -> getOrCreateStateSet() -> setMode(GL_DEPTH_TEST, osg::StateAttribute::ON);
    _shapes -> getOrCreateStateSet() -> setMode(GL_BLEND, osg::StateAttribute::ON);
    _shapes -> getOrCreateStateSet() -> setRenderingHint(osg::StateSet::TRANSPARENT_BIN);

    _viewer -> setThreadingModel(osgViewer::CompositeViewer::SingleThreaded);

    // This ensures that the widget will receive keyboard events. This focus
    // policy is not set by default. The default, Qt::NoFocus, will result in
    // keyboard events that are ignored.
    this -> setFocusPolicy(Qt::StrongFocus);
    this -> setMinimumSize(200, 200);

    // Ensures that the widget receives mouse move events even though no
    // mouse button has been pressed. We require this in order to let the
    // graphics window switch viewports properly.
    this -> setMouseTracking(true);
}

void Viewer::paintEvent(QPaintEvent * paintEvent)
{
    // this->makeCurrent();
    QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);
    paintGL();
    painter.end();
    swapBuffers();
    doneCurrent();
}

void
Viewer::paintGL()
{
    _viewer -> frame();
}

void Viewer::resizeGL(int width, int height)
{
  _graphics_window -> resized(0, 0, width, height);
    osgViewer::ViewerBase::Views views;
    _viewer->getViews(views);
    //for(osgViewer::View * view : views) { view -> resize(); }
}

osgViewer::View *
Viewer::get_view_with_focus()
{
   return _viewer -> getViewWithFocus();
}

void Viewer::keyPressEvent(QKeyEvent* event)
{
    QString keyString = event->text();
    const char* keyData = keyString.toLocal8Bit().data();
    // unsigned int view_index = _get_view_index_with_focus();
    // std::cout << "Index => " << view_index << std::endl;
    switch (event->key()) {
    case Qt::Key_Q:
        if (event->modifiers() & Qt::ControlModifier) {
            QApplication::quit();
        }
        // case Qt::Key_P          :   if(event->modifiers() & Qt::ControlModifier)
        //                             {
        //                                 QApplication::quit();
        //                             }
        // case Qt::Key_S          :   if(event->modifiers() & Qt::ShiftModifier)
        //                             {
        //                                 toggle_horizontal_split();
        //                             }
        // case Qt::Key_Z          :   if(event->modifiers() & Qt::ShiftModifier)
        //                             {
        //                                 toggle_vertical_split();
        //                             }
        // case Qt::Key_S          :   if(event->modifiers() & Qt::ShiftModifier)
        //                             {
        //                                 capture_continuous_toggle(view_index);
        //                             }
        //                             else
        //                             {
        //                                 capture_once(view_index);
        //                             }
        break;
    // case Qt::Key_Space:
    //     home();
    //     break;
    // case Qt::Key_Up:
    //     up(up_distance);
    //     break;
    // case Qt::Key_Down:
    //     down(down_distance);
    //     break;
    // case Qt::Key_Left:
    //     left(left_distance);
    //     break;
    // case Qt::Key_Right:
    //     right(right_distance);
    //     break;
    // case Qt::Key_F:
    //     forward(forward_distance);
    //     break;
    // case Qt::Key_B:
    //     backward(backward_distance);
    //     break;
    // case Qt::Key_Plus:
    //     zoom(zoom_factor);
    //     break;
    // case Qt::Key_Equal:
    //     zoom(zoom_factor);
    //     break;
    // case Qt::Key_Period:
    //     zoom(zoom_factor);
    //     break;
    // case Qt::Key_Greater:
    //     zoom(zoom_factor);
    //     break;
    // case Qt::Key_Minus:
    //     zoom(-zoom_factor);
    //     break;
    // case Qt::Key_Underscore:
    //     zoom(-zoom_factor);
    //     break;
    // case Qt::Key_Comma:
    //     zoom(-zoom_factor);
    //     break;
    // case Qt::Key_Less:
    //     zoom(-zoom_factor);
    //     break;
    // case Qt::Key_R:
    //     if (event->modifiers() & Qt::ShiftModifier) {
    //         roll(-roll_angle);
    //     }
    //     else {
    //         roll(roll_angle);
    //     }
    //     break;
    case Qt::Key_P:
        if (event->modifiers() & Qt::ControlModifier) {
            int x = 0, y = 0;
            unsigned int width = 1000, height = 500;
            // viewer.getCamera() -> getProjectionRectangle(x,y,width,height);
            osg::ref_ptr<osg::Image> image = new osg::Image;
            image->readPixels(x, y, width, height, GL_RGB, GL_UNSIGNED_BYTE);
            osgDB::writeImageFile(*image, "./image.jpg");
        }
    //     else if (event->modifiers() & Qt::ShiftModifier) {
    //         pitch(-pitch_angle);
    //     }
    //     else {
    //         pitch(pitch_angle);
    //     }
    //     break;
    // case Qt::Key_Y:
    //     if (event->modifiers() & Qt::ShiftModifier) {
    //         yaw(-yaw_angle);
    //     }
    //     else {
    //         yaw(yaw_angle);
    //     }
        break;
    default:
        break;
    }
}

// void
// Viewer::home()
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> home();
//     }
// }

// void
// Viewer::forward(double distance_delta)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> forward(distance_delta);
//     }
// }

// void
// Viewer::backward(double distance_delta)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> backward(distance_delta);
//     }
// }

// void
// Viewer::left(double distance_delta)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> left(distance_delta);
//     }
// }

// void
// Viewer::right(double distance_delta)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> right(distance_delta);
//     }
// }

// void
// Viewer::up(double distance_delta)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> up(distance_delta);
//     }
// }

// void
// Viewer::down(double distance_delta)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> down(distance_delta);
//     }
// }

// void
// Viewer::zoom(double factor)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> zoom(factor);
//     }
// }

// void
// Viewer::roll(double angle)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> roll(angle);
//     }
// }

// void
// Viewer::pitch(double angle)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> pitch(angle);
//     }
// }

// void
// Viewer::yaw(double angle)
// {
//     osgViewer::ViewerBase::Views views;
//     _viewer->getViews(views);
//     for(osgViewer::View * view : views)
//     {
//         view -> yaw(angle);
//     }
// }

void Viewer::keyReleaseEvent(QKeyEvent* event)
{
    QString keyString = event -> text();
    const char* keyData = keyString.toLocal8Bit().data();
    this->getEventQueue()->keyRelease(osgGA::GUIEventAdapter::KeySymbol(*keyData));
}

void Viewer::mouseMoveEvent(QMouseEvent* event)
{
    this->getEventQueue()->mouseMotion(static_cast<float>(event->x()),
        static_cast<float>(event->y()));
}

void Viewer::mousePressEvent(QMouseEvent* event)
{
    // 1 = left mouse button
    // 2 = middle mouse button
    // 3 = right mouse button

    unsigned int button = 0;

    switch (event->button()) {
    case Qt::LeftButton:
        button = 1;
        break;

    case Qt::MiddleButton:
        button = 2;
        break;

    case Qt::RightButton:
        button = 3;
        break;

    default:
        break;
    }

    this->getEventQueue()->mouseButtonPress(
        static_cast<float>(event->x()), static_cast<float>(event->y()), button);
}

void Viewer::mouseReleaseEvent(QMouseEvent* event)
{

    // 1 = left mouse button
    // 2 = middle mouse button
    // 3 = right mouse button

    unsigned int button = 0;

    switch (event->button()) {
    case Qt::LeftButton:
        button = 1;
        break;

    case Qt::MiddleButton:
        button = 2;
        break;

    case Qt::RightButton:
        button = 3;
        break;

    default:
        break;
    }

    this->getEventQueue()->mouseButtonRelease(
        static_cast<float>(event->x()), static_cast<float>(event->y()), button);
}

void Viewer::wheelEvent(QWheelEvent* event)
{

    event->accept();
    int delta = event->delta();

    osgGA::GUIEventAdapter::ScrollingMotion motion = delta > 0 ? osgGA::GUIEventAdapter::SCROLL_UP
                                                               : osgGA::GUIEventAdapter::SCROLL_DOWN;

    this->getEventQueue()->mouseScroll(motion);
}

bool Viewer::event(QEvent* event)
{
    bool handled = QGLWidget::event(event);

    // This ensures that the OSG widget is always going to be repainted after the
    // user performed some interaction. Doing this in the event handler ensures
    // that we don't forget about some event and prevents duplicate code.
    switch (event->type()) {
    case QEvent::KeyPress:
    case QEvent::KeyRelease:
    case QEvent::MouseButtonDblClick:
    case QEvent::MouseButtonPress:
    case QEvent::MouseButtonRelease:
    case QEvent::MouseMove:
    case QEvent::Wheel:
        this->update();
        break;
    default:
        break;
    }
    if (event->type() == QEvent::MouseButtonPress
        // && QApplication::keyboardModifiers() & Qt::ControlModifier
        )
    {
        osgViewer::View* view = _viewer->getView(0);
        osgUtil::LineSegmentIntersector::Intersections intersections;
        bool result = view->computeIntersections(
             view->getCamera(), osgUtil::Intersector::CoordinateFrame::PROJECTION,
             this->getEventQueue()->getCurrentEventState()->getXnormalized(),
             this->getEventQueue()->getCurrentEventState()->getYnormalized(),
             intersections);
         if (result) {
             const osgUtil::LineSegmentIntersector::Intersection& hit = *intersections.begin();
             this->getEventQueue()->mouseButtonRelease(
                 this->getEventQueue()->getCurrentEventState()->getX(),
                 this->getEventQueue()->getCurrentEventState()->getY(), 1);
             emit selected(hit.drawable->asGeometry()->getName().c_str());
          }
    }
    return (handled);
}


osgGA::EventQueue* Viewer::getEventQueue() const
{
    osgGA::EventQueue* event_queue = _graphics_window->getEventQueue();

    if (event_queue)
        return (event_queue);
    else
        throw(std::runtime_error("Unable to obtain valid event queue"));
}

void
Viewer::attach_view(MeshView * mesh_view)
{
    mesh_view -> set_graphics_window(_graphics_window.get());
    mesh_view -> get_view() -> setSceneData(_shapes.get());
    _viewer -> addView(mesh_view -> get_view());
    mesh_view -> home();
}

void
Viewer::detach_view(MeshView * mesh_view)
{
  _viewer -> removeView(mesh_view -> get_view());
}

void
Viewer::attach_shape(MoogliShape * shape)
{
    _shapes -> addChild(shape -> root.get());
}

void
Viewer::detach_shape(MoogliShape * shape)
{
    _shapes -> removeChild(shape -> root.get());
}

File: moogli/core/viewer.py
import PyQt4
from PyQt4 import QtCore, QtGui
from PyQt4.QtCore import Qt
import math
import _moogli

class View(_moogli.MeshView):
    def __init__(self, id):
        self._id = id
        _moogli.MeshView.__init__(self)
        self._color_bars = set()

    def get_id(self):
        return self._id

    def attach_color_bar(self, color_bar):
        if color_bar in self._color_bars: return
        color_bar.view = self
        super(View, self).attach_color_bar(color_bar)
        self._color_bars.add(color_bar)

    def resize(self, width, height):
        super(View, self).resize(width, height)
        for color_bar in self._color_bars:
            color_bar.resize(width, height)



class Viewer(_moogli.Viewer):
    def __init__( self
                # , parent
                # , shareWidget
                # , flags
                , prelude   = lambda x : x
                , interlude = lambda x : x
                , postlude  = lambda x : x
                , idletime  = 0
                ):
        PyQt4.QtGui.QApplication.instance().aboutToQuit.connect(self.stop)
        _moogli.Viewer.__init__(self)
        self.groups     = set()
        # self.network    = network
        self.prelude    = prelude
        self.interlude  = interlude
        self.postlude   = postlude
        self.idletime   = idletime
        self._done      = False
        self._pause     = False
        self._views = dict()
        self._meshes = set()
        self.roll_angle = math.pi / 36.0
        self.pitch_angle = math.pi / 36.0
        self.yaw_angle = math.pi / 36.0
        self.zoom_factor = 0.10
        self.forward_distance = 10.0
        self.backward_distance = 10.0
        self.left_distance = 10.0
        self.right_distance = 10.0
        self.up_distance = 10.0
        self.down_distance = 10.0

    def start(self):
        self._done = False
        self.prelude(self)
        self.update()
        self.run()

    def pause(self):
        self._pause = True

    def resume(self):
        self._pause = False
        self.run()

    def stop(self):
        self._done = True
        self.postlude(self)

    def run(self):
        self.interlude(self)
        self.update()
        if self._done: return
        if self._pause: return
        PyQt4.QtCore.QTimer.singleShot(self.idletime, self.run)

    def resizeEvent(self, event):
        super(Viewer, self).resizeEvent(event)
        for view in self._views.values():
            view.resize(event.size().width(), event.size().height())

    def attach_view(self, view):
        if view in self._views.values(): return
        super(Viewer, self).attach_view(view)
        self._views[view.get_id()] = view

    def detach_view(self, view):
        if view not in self._views.values(): return
        super(Viewer, self).detach_view(view)
        self._views.pop(view.get_id())

    def attach_geometry(self, mesh):
        if mesh in self._meshes: return
        super(Viewer, self).attach_geometry(mesh)
        self._meshes.add(mesh)

    def home(self):
        [view.home() for view in self._views.values()]

    def forward(self, distance_delta):
        [view.forward(distance_delta) for view in self._views.values()]

    def backward(self, distance_delta):
        [view.backward(distance_delta) for view in self._views.values()]

    def left(self, distance_delta):
        [view.left(distance_delta) for view in self._views.values()]

    def right(self, distance_delta):
        [view.right(distance_delta) for view in self._views.values()]

    def up(self, distance_delta):
        [view.up(distance_delta) for view in self._views.values()]

    def down(self, distance_delta):
        [view.down(distance_delta) for view in self._views.values()]

    def zoom(self, factor):
        [view.zoom(factor) for view in self._views.values()]

    def roll(self, angle):
        [view.roll(angle) for view in self._views.values()]

    def pitch(self, angle):
        [view.pitch(angle) for view in self._views.values()]

    def yaw(self, angle):
        [view.yaw(angle) for view in self._views.values()]


    def keyPressEvent(self, event):
        #super(Viewer, self).event(event)
        key = event.key()
        # modifiers = QtGui.QApplication.keyboardModifiers()
        control_pressed = event.modifiers() & Qt.ControlModifier == 1
        shift_pressed = event.modifiers() & Qt.ShiftModifier == 1
        if key == Qt.Key_Q and control_pressed:
            QApplication.quit()
        elif key == Qt.Key_W and control_pressed:
            QApplication.quit()
        elif key == Qt.Key_Space:
            self.home()
        elif key == Qt.Key_Up:
            self.up(self.up_distance)
        elif key == Qt.Key_Down:
            self.down(self.down_distance)
        elif key == Qt.Key_Left:
            self.left(self.left_distance)
        elif key == Qt.Key_Right:
            self.right(self.right_distance)
        elif key == Qt.Key_F:
            self.forward(self.forward_distance)
        elif key == Qt.Key_B:
            self.backward(self.backward_distance)
        elif key == Qt.Key_Plus:
            self.zoom(self.zoom_factor)
        elif key == Qt.Key_Equal:
            self.zoom(self.zoom_factor)
        elif key == Qt.Key_Period:
            self.zoom(self.zoom_factor)
        elif key == Qt.Key_Greater:
            self.zoom(self.zoom_factor)
        elif key == Qt.Key_Minus:
            self.zoom(-self.zoom_factor)
        elif key == Qt.Key_Underscore:
            self.zoom(-self.zoom_factor)
        elif key == Qt.Key_Comma:
            self.zoom(-self.zoom_factor)
        elif key == Qt.Key_Less:
            self.zoom(-self.zoom_factor)
        elif key == Qt.Key_R:
            if shift_pressed:
                self.roll(-self.roll_angle)
            else:
                self.roll(self.roll_angle)
        elif key == Qt.Key_P:
            if shift_pressed:
                self.pitch(-self.pitch_angle)
            else:
                self.pitch(self.pitch_angle)
        elif key == Qt.Key_Y:
            if shift_pressed:
                self.yaw(-self.yaw_angle)
            else:
                self.yaw(self.yaw_angle)

    def get_view_with_focus(self):
        view_id = super(Viewer, self).get_view_with_focus()
        return self._views[view_id]

    @QtCore.pyqtSlot(str)
    def printme(self, name):
        # cursor = QtGui.QCursor().pos()
        menu = QtGui.QMenu()
        menu.addAction("a")
        menu.addAction("b")
        menu.addAction("c")
        menu.exec_(QtGui.QCursor().pos())
        print("Selected = > ", name)
